src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,296 @@
+                }
+
+                self.action(input, input_proj);
+                *relation = input.take_dangerous().filter(predicates.clone()).project(
+                    projection
+                        .into_iter()
+                        .map(|c| *col_map.get(&c).unwrap())
+                        .collect(),
+                );
+            }
+            // TODO(justin): can the arms for FlatMap and Map be unified somehow?
+            RelationExpr::FlatMap {
+                input, func, exprs, ..
+            } => {
+                let input_arity = input.arity();
+                let mut needed: HashSet<_> = projection.iter().cloned().collect();
+                // First, prune any elements of scalars that are ignored by the projection.
+                // Compute the transitive closure of the needed columns.
+                for expr in exprs.iter() {
+                    needed.extend(expr.support());
+                }
+                // First, prune away unneeded columns from the input.
+                let mut col_map = HashMap::new();
+                let mut input_proj = vec![];
+                for i in 0..input_arity {
+                    if needed.contains(&i) {
+                        col_map.insert(i, input_proj.len());
+                        input_proj.push(i);
+                    }
+                }
+
+                for expr in exprs.iter_mut() {
+                    expr.visit_mut(&mut |e| {
+                        if let ScalarExpr::Column(c) = e {
+                            *c = *col_map.get(c).unwrap();
+                        }
+                    });
+                }
+                for c in projection.iter_mut() {
+                    if *c < input_arity {
+                        *c = *col_map.get(c).unwrap();
+                    } else {
+                        *c = *c - input_arity + input_proj.len();
+                    }
+                }
+                self.action(input, input_proj);
+                *relation = input
+                    .take_dangerous()
+                    .flat_map(func.clone(), exprs.clone())
+                    .project(projection);
+            }
+            RelationExpr::Map { input, scalars } => {
+                let input_arity = input.arity();
+                let mut needed: HashSet<_> = projection.iter().cloned().collect();
+                // First, prune any elements of scalars that are ignored by the projection.
+                // Compute the transitive closure of the needed columns.
+                let needed = loop {
+                    let mut new = needed.clone();
+                    for c in needed.iter() {
+                        if *c >= input_arity {
+                            new.extend(scalars[*c - input_arity].support());
+                        }
+                    }
+                    if new == needed {
+                        break new;
+                    }
+                    needed = new;
+                };
```

Comment:
A different approach with (for me) clearer structure is to iterate through `(0 .. scalars.len()).rev()`, testing each to see if it is needed and incorporating its dependencies if so. I think the above has the potential to be quadratic in `scalars.len()`, which I think this proposal would also avoid.